rllib/core/models/catalog.py

```diff
@@ -0,0 +1,227 @@
+
+
+class Catalog:
+    """Defines how what models an RLModules builds.
+
+    RLlib's native RLModules get their Models from a Catalog object.
+    By default, that Catalog builds the configs it holds.
+    You can modify a Catalog so that it builds different Models by subclassing and
+    overriding the build_* methods. Alternatively, you can customize the configs
+    inside RLlib's Catalogs to customize what is being built by RLlib.
+
+    Usage example:
+
+    # Define a custom catalog
+
+    .. testcode::
+
+    class MyCatalog(Catalog):
+    def __init__(
+        self,
+        observation_space: gym.Space,
+        action_space: gym.Space,
+        model_config_dict: dict,
+    ):
+        super().__init__(observation_space, action_space, model_config_dict)
+        self.my_model_config = MLPModelConfig(
+            hidden_layer_dims=[64, 32],
+            input_dim=self.observation_space.shape[0],
+            output_dim=1,
+        )
+
+        def build_my_head(self, framework: str):
+            return self.my_model_config.build(framework=framework)
+
+
+    # With that, RLlib can build and use models from this catalog like so:
+    catalog = MyCatalog(gym.spaces.Box(0, 1), gym.spaces.Box(0, 1), {})
+    my_head = catalog.build_my_head("torch")  # doctest: +SKIP
+    out = my_head(...)  # doctest: +SKIP
+
+    # We can also modify configs of RLlib's native Catalogs like so:
+    catalog = MyCatalog(gym.spaces.Box(0, 1), gym.spaces.Box(0, 1), {})
+    catalog.my_model_config.output_dim = 32
+    my_head = catalog.build_my_head("torch")  # doctest: +SKIP
+    out = my_head(...)  # doctest: +SKIP
+
+    """
+
+    def __init__(
+        self,
+        observation_space: gym.Space,
+        action_space: gym.Space,
+        # TODO (Artur): Turn model_config into model_config_dict to distinguish
+        #  between ModelConfig and a model_config dict.
+        model_config: dict,
+        view_requirements: dict = None,
+    ):
+        """Initializes a Catalog with a default encoder config.
+
+        Args:
+            observation_space: The observation space of the environment.
+            action_space: The action space of the environment.
+            model_config: The model config that specifies things like hidden
+                dimensions and activations functions to use in this Catalog.
+            view_requirements: The view requirements of Models to produce. This is
+                needed for a Model that encodes something else than observations or
+                actions but not otherwise.
+
+        """
+        self.observation_space = observation_space
+        self.action_space = action_space
+
+        # TODO (Artur): Possibly get rid of this config merge
+        self.model_config = {**MODEL_DEFAULTS, **model_config}
+        self.view_requirements = view_requirements
+
+        # Produce a basic encoder config.
+        self.encoder_config = self.get_encoder_config(
```

Comment:
I don't think it should be protected. When writing your own catalog, this would be a valid place to add extra logic.
Also, it does not change any instance variables of the Catalog, so it is safe to query a catalog for Encoders with it and see how it behaves, how to produce certain types of encodes etc. Especially if we add more complex models, this would be a good place (next to the docs) for users to explore and understand this catalog API. I wouldn't want to deter them from it.

Reply:
The constructor args are there because they are used for creating LSTM cells.
LSTM cells potentially need view_requirements and the action space. The other args are obvious I guess.

Reply:
> The constructor args are there because they are used for creating LSTM cells.
LSTM cells potentially need view_requirements and the action space. The other args are obvious I guess.

for args. I mean why don't you use self.

Reply:
> I don't think it should be protected. When writing your own catalog, this would be a valid place to add extra logic.

By protected I mean a method that can be overriden but is not assumed to be public.

Reply:
I think by making it a class method the discussion is killed :)